src/HomeNetProtocolTests/Tests/HN05025.cs

```diff
@@ -0,0 +1,408 @@
+        string calleeMessage = "Invalid Message";
+        messageBytes = Encoding.UTF8.GetBytes(calleeMessage);
+        requestMessageAppServiceCallee = mbCallee1AppService.CreateApplicationServiceSendMessageRequest(calleeToken2, messageBytes);
+        await clientCallee1AppService.SendMessageAsync(requestMessageAppServiceCallee);
+
+        Message responseAppServiceCallee = await clientCallee1AppService.ReceiveMessageAsync();
+        idOk = responseAppServiceCallee.Id == requestMessageAppServiceCallee.Id;
+        statusOk = responseAppServiceCallee.Response.Status == Status.ErrorNotFound;
+
+        bool sendMessageOk = idOk && statusOk;
+
+
+        // Step 7 Acceptance
+        bool step7Ok = sendMessageOk;
+
+        log.Trace("Step 7: {0}", step7Ok ? "PASSED" : "FAILED");
+
+
+        // Step 8 
+        log.Trace("Step 8");
+        await Task.Delay(3000);
+        string caller2Message1 = "Message #1 to callee from caller2.";
+        messageBytes = Encoding.UTF8.GetBytes(caller2Message1);
+        requestMessageAppServiceCaller = mbCaller2AppService.CreateApplicationServiceSendMessageRequest(callerToken2, messageBytes);
+
+        // Either the third client is disconnected and this should prevent sending a message or receiving a response,
+        // OR just the relay was destroyed and the client will receive error not found.
+        bool disconnectOk = false;
+        messageOk = false;
+        try
+        {
+          await clientCaller2AppService.SendMessageAsync(requestMessageAppServiceCaller);
+          responseMessage = await clientCaller2AppService.ReceiveMessageAsync();
+          idOk = responseMessage.Id == requestMessageAppServiceCaller.Id;
+          statusOk = responseMessage.Response.Status == Status.ErrorNotFound;
+          messageOk = idOk && statusOk;
+        }
+        catch
+        {
+          log.Trace("Expected exception occurred.");
+          disconnectOk = true;
+        }
+
+        // Step 8 Acceptance
+        bool step8Ok = disconnectOk || messageOk;
+
+        log.Trace("Step 8: {0}", step8Ok ? "PASSED" : "FAILED");
+
+
+        // Step 9
+        log.Trace("Step 9");
+
+        // Receive ACK message #1.
+        responseMessageAppServiceCaller = await clientCaller1AppService.ReceiveMessageAsync();
+        idOk = responseMessageAppServiceCaller.Id == callerMessage1Id;
+        statusOk = responseMessageAppServiceCaller.Response.Status == Status.Ok;
+        receivedVersion = responseMessageAppServiceCaller.Response.SingleResponse.Version.ToByteArray();
+        versionOk = StructuralComparisons.StructuralComparer.Compare(receivedVersion, new byte[] { 1, 0, 0 }) == 0;
+
+        bool receiveAckOk = idOk && statusOk && versionOk;
+
+        // Step 9 Acceptance
+        bool step9Ok = receiveAckOk;
+
+        log.Trace("Step 9: {0}", step9Ok ? "PASSED" : "FAILED");
+
+        Passed = step1Ok && step2Ok && step3Ok && step4Ok && step5Ok && step6Ok && step7Ok && step8Ok && step9Ok;

```

Comment:
I got the task to review 10 000 lines of code every week, but you make that quite hard by not learning how to communicate through code. You write this code for

- The compiler to create an application from it
- Yourself in about 2 months so that you can read it back and change it
- Other developers, so they can follow your train of thoughts so they can learn from it or God forbid fix something you might miss.

This method only serves the 1st purpose and even you will not be able to change this test with names like "step 8" instead of a descriptive name. Yes, I can find out what step 8 does *after* I read the implementation. The important thing in giving names is to parts of the code is to hide the details and provide a "table of contents" to the reader so they can navigate without reading the whole implementation.